### PR TITLE
[GraphOptz] Fold Exp+ReduceSum+Div into Softmax

### LINKED ISSUE #4654: [GraphOptz] Fold Exp + ReduceSum + Div into Softmax
Some ONNX models contain this:
image
This should be folded into Softmax. The most important reason for this is the numerical stability since Exp, Div are numerically unstable whereas the Softmax is specially implemented/handled to avoid high values of the Exp.

### PR DESCRIPTION
**Pull request by Squadrick**
Summary: Optimize exp + reduce sum + div into a single softmax node. This provides extra numerical stability as well.

Fixes #4654 

Test Plan: A test has been added to `GraphOptzTest` checking the optimization for a graph with the above three nodes, and the optimized graph only has a single softmax.

### CODE REVIEW

**Comment by Squadrick** [comment]
cc @jfix71 request for review. The failed test looks unrelated: 

```
The following tests FAILED:
	  6 - ThreadPoolExecutorTest (Failed)
Errors while running CTest
FAILED: CMakeFiles/check 
cd /var/lib/jenkins/workspace/build && /usr/local/lib/python2.7/dist-packages/cmake/data/bin/ctest -LE EXPENSIVE\|STRESS
ninja: build stopped: subcommand failed.
```

**Review by 842974287 — COMMENTED** [review]
Hi @Squadrick, thanks for adding this! Overall looks pretty good. I have a question that since we are folding these node for extra numerical stability, should we still do the folding even ExpNode has other users? What you thought on this

**Comment by Squadrick** [comment]
@842974287 It'll make the logic a bit more complex, consider:

```
    IN
     |
    Exp (Node 1)
   /   \ 
  |  ReduceSum (Node 2)
   \   /
    Div
     |
    OUT
```

If we have users of nodes `1` or `2`, we'll need to make copies of these nodes and manually replace them in the model. I don't see a lot of cases where these intermediate nodes will be used, especially something using `2`. Adding support for additional users of `EXP (Node 1)` should be useful.  

To illustrate:

```
    IN
     |
    Exp -- Other nodes
   /   \ 
  |  ReduceSum
   \   /
    Div
     |
    OUT
```
will become:
```
     IN
   /    \
Softmax  EXP -- Other nodes
  |
 OUT
```

If this looks good, I'll go ahead with the implementation.

**Comment by 842974287** [comment]
@Squadrick 

> If we have users of nodes 1 or 2, we'll need to make copies of these nodes and manually replace them in the model.

I'm not sure I'm following here. Why would we need to make copies and manually replace here? We can just replace all the users of Div with Softmax and the rest should be fine. I'm also very new to Glow so please correct me if I'm wrong.

> If this looks good, I'll go ahead with the implementation.

Yeah this looks good! If we stop at DivNode and check its inputs, this should make the logic a little bit easier.

File: include/glow/Optimizer/GraphOptimizer/FunctionPasses.def [hunk 1]
```diff
@@ -60,6 +60,7 @@ FUN_PASS(OptimizeConcatQuantization)
 FUN_PASS(SinkConcatBelowQuantize)
 FUN_PASS(FoldLayerNormArithmetic)
 FUN_PASS(QuantizeSwish)
+FUN_PASS(OptimizeExpSumDiv)
```
Comment (jfix71) [inline include/glow/Optimizer/GraphOptimizer/FunctionPasses.def:63]: nit: Probably makes sense to call this `FoldExpSumDiv` or similar

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp [hunk 2]
```diff
@@ -5088,6 +5088,76 @@ bool QuantizeSwish::run(Function *F, const CompilationContext &cctx) {
...
+
+  bool changed = false;
+  for (auto &N : F->getNodes()) {
+    auto *EN = dyn_cast<ExpNode>(&N);
+    if (!EN || EN->getNumUsers() != 2) {
+      continue;
+    }
+
+    DivNode *DN;
+    BatchedReduceAddNode *RSN;
```
Comment (jfix71) [inline lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:5112]: nit: Initialize these to nullptr

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp [hunk 3]
```diff
@@ -5088,6 +5088,76 @@ bool QuantizeSwish::run(Function *F, const CompilationContext &cctx) {
...
+    auto *user2 = EN->getUsers().back().getUser();
+
+    if (isa<DivNode>(user1) && isa<BatchedReduceAddNode>(user2)) {
+      DN = cast<DivNode>(user1);
+      RSN = cast<BatchedReduceAddNode>(user2);
+    } else if (isa<DivNode>(user2) && isa<BatchedReduceAddNode>(user1)) {
+      DN = cast<DivNode>(user2);
+      RSN = cast<BatchedReduceAddNode>(user1);
+    } else {
+      optimizationPossible = false;
```
Comment (jfix71) [inline lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:5125]: nit: Could just continue here, and get rid of `optimizationPossible` entirely.

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp [hunk 4]
```diff
@@ -5088,6 +5088,76 @@ bool QuantizeSwish::run(Function *F, const CompilationContext &cctx) {
...
+    }
+
+    if (!optimizationPossible || RSN->getNumUsers() != 1) {
+      continue;
+    }
+
+    // DivNode is a broadcasted node when introduced two extra nodes
+    // for broadcasting and tiling. Ensure that these nodes exist in the
```
Comment (jfix71) [inline lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:5133]: nit: Grammar here on the first sentence is a bit hard to parse. Is it supposed to be something like "when it introduces"?

File: lib/Optimizer/GraphOptimizerPipeline/FunctionPassPipeline.cpp [hunk 5]
```diff
@@ -141,6 +141,9 @@ createDefaultGraphOptimizationPassPipeline() {
       // Optimize reshapes introduced during above optimizations.
       {FunctionPassID::OptimizeReshape},
 
+      // Optimize exp + reduce sum + div into softmax
+      {FunctionPassID::OptimizeExpSumDiv},
```
Comment (jfix71) [inline lib/Optimizer/GraphOptimizerPipeline/FunctionPassPipeline.cpp:145]: I think we should add this to `createDefaultFoldPassPipeline` instead.

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp [hunk 6]
```diff
@@ -5088,6 +5088,76 @@ bool QuantizeSwish::run(Function *F, const CompilationContext &cctx) {
...
+    // graph.
+    auto *broadcastNode = getOnlyUser(*RSN);
+    if (broadcastNode == nullptr) {
+      continue;
+    }
+    auto *tileNode = getOnlyUser(*broadcastNode);
+    if (tileNode == nullptr) {
+      continue;
+    }
```
Comment (jfix71) [inline lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:5142]: After you put up this PR an explicit BroadcastNode was added (lowered to Reshape/Tile) in https://github.com/pytorch/glow/pull/5047. Since this pass will occur before lowering, I believe we want to just look for the BroadcastNode instead of the Tile. (If you rebase I imagine your graph opt test will fail)

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp [hunk 7]
```diff
@@ -5088,6 +5088,76 @@ bool QuantizeSwish::run(Function *F, const CompilationContext &cctx) {
...
+    auto *tempDN = getOnlyUser(*tileNode);
```
[the remark this answers is not in the document]
Comment (jfix71) [reply lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:5143]: I think it's sane to keep.

File: tests/unittests/GraphOptzTest.cpp [hunk 8]
```diff
@@ -6080,3 +6080,31 @@ TEST_F(GraphOptz, lower1x1AdaptiveAvgPoolToAvgPool) {
...
+
+  EXPECT_EQ(2, optimizedF_->getNodes().size());
+
+  EXPECT_EQ(0, countNodeKind(optimizedF_, Kinded::Kind::ExpNodeKind));
+  EXPECT_EQ(0, countNodeKind(optimizedF_, Kinded::Kind::DivNodeKind));
+  EXPECT_EQ(0,
+            countNodeKind(optimizedF_, Kinded::Kind::BatchedReduceAddNodeKind));
+  EXPECT_EQ(1, countNodeKind(optimizedF_, Kinded::Kind::SoftMaxNodeKind));
+
+  checkNumericalEquivalence(1e-7f);
```
Comment (jfix71) [inline tests/unittests/GraphOptzTest.cpp:6109]: Just curious if you were seeing actual numerical differences here, or if we could change this to `0.f` on the allowed error.

**Review by jfix71 — COMMENTED** [review]
Thanks for adding this @Squadrick! Sorry for the delayed review.

File: tests/unittests/GraphOptzTest.cpp (hunk 8 above)
Comment (Squadrick) [reply tests/unittests/GraphOptzTest.cpp:6109]: Changing it to `0.f` causes the test to fail.
